README.md
# LXC-Web-Panel

This is a fork of LXC-Web-Panel from https://github.com/lxc-webpanel/LXC-Web-Panel

Tested on Ubuntu 12.04 using lxc from daily ppa: https://launchpad.net/~ubuntu-lxc/+archive/daily

## Installation

_Deprecated!!_
```
wget https://raw2.github.com/claudyus/LXC-Web-Panel/master/tools/install.sh -O - | sudo bash
```

## Configuration

1. Copy lwp.example.conf as /etc/lwp/lwp.conf
2. edit it
3. restart lwp service

### SSL configuration
To add SSL Support add to global section of lwp.conf this entries:

```
ssl = True
pkey = mykey.key
cert = mykey.cert
```

Where mykey.key and mykey.cert are the key and the certificate generated previously for example by the command:
```
 openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout mykey.key -out mykey.cert
```

## Update

```
cd /srv/lwp
sudo git pull
```
Than check the lwp.example.conf with your /etc/lwp/lwp.conf file for new options and depends.

## Remove
```
sudo rm -fr /srv/lwp
sudo rm /etc/init.d/lwp
sudo update-rc.d lwp remove

```

### Info
This repo contains a lot of mixup from various forks, see https://github.com/claudyus/LXC-Web-Panel/network for details.


CHANGELOG
0.6 / 2014-07-08
================

  fix: parse all values of lxc-info
  Update README and specify dependencies version
  fix: parse all values of lxc-info
  Update README and specify dependencies version
  frontend: add a view to manage the api token, table api_token modified
  fixup: api GET containers return a more useful information
  fixup; return json error if 401, use header to auth
  api: implement management token apis, new table api_tokens(user_id int, description varchar(30), token varchar(50) not null)
  api: working auth api decorator
  api: more code draft
  api: code draft
  api: update api draft, remove X- prefix due to RFC6648
  api: spec draft
  fix: pid assignment from lxc-info
  save version in debian package, read it
  delete gh-pages as submodules
  fix: ldap use the correct user base from search query
  rework: remove useless html from buckets branch

0.5 / 2014-05-14
================

  feat: notify the user if ldap auth is used
  doc: update README and conf example
  fix: README typo
  fixup: moer code on push bucket
  feat: initial develop for file-bucket integration
  Merge branch 'debian'
  debian: fix lwp initscript add prerm script
  fix: lwp example use database inside /srv/lwp by default
  fix: some details for debian package, postinst doesn't work...
  fix Readme syntax
  integrate gh-pages, package build, README to rst format
  debian: add debian package
  fix: lint the code of lxclite class
  feat: allow override of debug option from command-line
  fix: lint the code with flake8 rules
  fixup: avoid to display flash message when autostart is unchanged
  fix: markdown with love...
  fix autostart problem, partially resolve issues #3
  Readme: add ssl config info, cleanup
  feat: move config file to /etc/lwp, remove update script, explain config option, deprecate install script
  fix: decode utf8 name from ldap server
  Initial support for zfs backing store.  Fixes #55
  Hostname will now be printed
  Added 14.04 to supported_dists
  Document the new configuration process
  feat: implement LDAP based auth as backend
  Code cleanup
  feat: display clone and backup buttons inside container edit page
  feat: dynamically load contributors from github API
  refactor: move css on separate file, prettify css

0.4 / 2014-04-14
================

  feat(backup): omit /var/lib/lxc prefix during backup, don't use ':' inside filename
  Update Readme.md with remove info
  Show container name in confirmation dialog when destroying a container.
  Show memory limit (if set) beside memory usage in overview page.
  Add .editorconfig
  feat: retrieve project version using git describe
  rework: use logge_in decorator for backup_container()
  fix: avoid exception during backup if container was stopped


0.3 / 2014-04-10
================

  delete compiled/ignored file
  fix lxc-clone for lxc v1.0
  fix lxc-info output parse
  Just some info on README, clone this repo during installation
  fix syntax after merge
  Merge remote-tracking branch 'xbddc/master'
  Merge remote-tracking branch 'tonnydourado/dev'
  Revert "Get IP addres directly from running container instead of config file"
  Merge remote-tracking branch 'rasmuskr/master'
  Merge remote-tracking branch 'nevelis/aptimize'
  Adding runtime files to gitignore
  Adding install and update scripts and updateing urls to point to this reposetory
  Merge branch 'master' of git://github.com/jureqq/LXC-Web-Panel
  added gitignore file
  Freezing container before backup and unfreeze after
  Using Tar and Gzip for backup Append date and time to the backup archive file name
  merge ssl and backup changes in lwp.conf
  fix merge
  Merge branch 'lwp-backup' of https://github.com/vaytess/LXC-Web-Panel
  Merge branch 'ssl' of https://github.com/remixtj/LXC-Web-Panel
  Get IP addres directly from running container instead of config file
  * fix platform info. * exclude file from containers list
  Merge pull request #51 from orgoj/patch-1
  Update __init__.py
  Adding .gitignore
  Display IPv4 addresses for containers using DHCP
  Added Ubuntu 13.10 to supported distributions
  Use correct variable for extra arguments
  Don't include files in container listing
  Merge pull request #44 from gweis/master
  make lxc web panel deployable behind reverse proxy by using url_for for all url’s in templates
  fixing lxc-info
  code refactoring within lxc_net() action
  fixing bug of 'internal server error (error 500)' when clicking to Networking link (sidebar nav) when the host is not ubuntu
  add backup_container() action in lwp.py and modal_backup.html view to templates/includes/
  add some params to lwp.config and add a backup() method in lxclite module
  newline missing
  Added SSL Support to lwp panel
  write container's ip back to config to speed up the load time of dashboard
  container can only be destroyed in stopped stage.
  detect container's ip address automatically when not specified in config file.
  Fixing argument name in if_logged_in decorator
  Ignoring the config file.
  Some refactoring on the lxc_net function.
  Added .gitignore file
  Added @if_logged_in decorator, plus some minor style changes
  Merge pull request #13 from 4lexx/master
  allow IP address in CIDR notation

0.2 / 2013-04-30
================

+ Instant push memory and CPU options to cgroup values
+ Added partition option in the config file, you can choose which partition you want to monitor in the overview
+ If there is no container, the panel will display the modal "Create CT"
+ Added clone CT feature
+ Added "start at boot" button
+ Added support for multiple key/value (only for lxc.cgroup.devices.deny and lxc.cgroup.devices.allow) in container's config file (with ConfigParser)
+ Added support for templates option (Create CT)

* Now, in the sidebar, links are highlighted for containers edit pages
* Added modal to confirm the deletion of a user
  * Can't delete a user if he is the only one
  * Can't delete a admin user if there is no other admin user

lwp/__init__.py
import sys
sys.path.append('../')
from lxclite import exists, stopped
import subprocess
import os
import platform
import time
import urllib2
import ConfigParser
import re


class CalledProcessError(Exception):
    pass


class LxcConfigFileNotComplete(Exception):
    pass

cgroup = {}
cgroup['type'] = 'lxc.network.type'
cgroup['link'] = 'lxc.network.link'
cgroup['flags'] = 'lxc.network.flags'
cgroup['hwaddr'] = 'lxc.network.hwaddr'
cgroup['rootfs'] = 'lxc.rootfs'
cgroup['utsname'] = 'lxc.utsname'
cgroup['arch'] = 'lxc.arch'
cgroup['ipv4'] = 'lxc.network.ipv4'
cgroup['memlimit'] = 'lxc.cgroup.memory.limit_in_bytes'
cgroup['swlimit'] = 'lxc.cgroup.memory.memsw.limit_in_bytes'
cgroup['cpus'] = 'lxc.cgroup.cpuset.cpus'
cgroup['shares'] = 'lxc.cgroup.cpu.shares'
cgroup['deny'] = 'lxc.cgroup.devices.deny'
cgroup['allow'] = 'lxc.cgroup.devices.allow'
cgroup['auto'] = 'lxc.start.auto'


class FakeSection(object):
    def __init__(self, fp):
        self.fp = fp
        self.sechead = '[DEFAULT]\n'

    def readline(self):
        if self.sechead:
            try:
                return self.sechead
            finally:
                self.sechead = None
        else:
            return self.fp.readline()


def DelSection(filename=None):
    if filename:
        load = open(filename, 'r')
        read = load.readlines()
        load.close()
        i = 0
        while i < len(read):
            if '[DEFAULT]' in read[i]:
                del read[i]
                break
        load = open(filename, 'w')
        load.writelines(read)
        load.close()


def file_exist(filename):
    '''
    checks if a given file exist or not
    '''
    try:
        with open(filename) as f:
            f.close()
            return True
    except IOError:
        return False


def memory_usage(name):
    '''
    returns memory usage in MB
    '''
    if not exists(name):
        raise ContainerNotExists("The container (%s) does not exist!" % name)
    if name in stopped():
        return 0
    cmd = ['lxc-cgroup -n %s memory.usage_in_bytes' % name]
    try:
        out = subprocess.check_output(cmd, shell=True).splitlines()
    except:
        return 0
    return int(out[0]) / 1024 / 1024


def host_memory_usage():
    '''
    returns a dict of host memory usage values
                    {'percent': int((used/total)*100),
                    'percent_cached':int((cached/total)*100),
                    'used': int(used/1024),
                    'total': int(total/1024)}
    '''
    out = open('/proc/meminfo')
    for line in out:
        if 'MemTotal:' == line.split()[0]:
            split = line.split()
            total = float(split[1])
        if 'MemFree:' == line.split()[0]:
            split = line.split()
            free = float(split[1])
        if 'Buffers:' == line.split()[0]:
            split = line.split()
            buffers = float(split[1])
        if 'Cached:' == line.split()[0]:
            split = line.split()
            cached = float(split[1])
    out.close()
    used = (total - (free + buffers + cached))
    return {'percent': int((used / total) * 100),
            'percent_cached': int(((cached) / total) * 100),
            'used': int(used / 1024),
            'total': int(total / 1024)}


def host_cpu_percent():
    '''
    returns CPU usage in percent
    '''
    f = open('/proc/stat', 'r')
    line = f.readlines()[0]
    data = line.split()
    previdle = float(data[4])
    prevtotal = float(data[1]) + float(data[2]) + float(data[3]) + float(data[4])
    f.close()
    time.sleep(0.1)
    f = open('/proc/stat', 'r')
    line = f.readlines()[0]
    data = line.split()
    idle = float(data[4])
    total = float(data[1]) + float(data[2]) + float(data[3]) + float(data[4])
    f.close()
    intervaltotal = total - prevtotal
    percent = 100 * (intervaltotal - (idle - previdle)) / intervaltotal
    return str('%.1f' % percent)


def host_disk_usage(partition=None):
    '''
    returns a dict of disk usage values
                    {'total': usage[1],
                    'used': usage[2],
                    'free': usage[3],
                    'percent': usage[4]}
    '''
    if not partition:
        partition = '/'
    usage = subprocess.check_output(['df -h %s' % partition], shell=True).split('\n')[1].split()
    return {'total': usage[1],
            'used': usage[2],
            'free': usage[3],
            'percent': usage[4]}


def host_uptime():
    '''
    returns a dict of the system uptime
            {'day': days,
            'time': '%d:%02d' % (hours,minutes)}
    '''
    f = open('/proc/uptime')
    uptime = int(f.readlines()[0].split('.')[0])
    minutes = uptime / 60 % 60
    hours = uptime / 60 / 60 % 24
    days = uptime / 60 / 60 / 24
    f.close()
    return {'day': days,
            'time': '%d:%02d' % (hours, minutes)}


def check_ubuntu():
    '''
    return the System version
    '''
    dist = '%s %s' % (platform.linux_distribution()[0], platform.linux_distribution()[1])

    supported_dists = [ 'Ubuntu 12.04',
                        'Ubuntu 12.10',
                        'Ubuntu 13.04',
                        'Ubuntu 13.10',
                        'Ubuntu 14.04']

    if dist in supported_dists:
        return dist
    return 'unknown'


def get_templates_list():
    '''
    returns a sorted lxc templates list
    '''
    templates = []
    path = None

    try:
        path = os.listdir('/usr/share/lxc/templates')
    except:
        path = os.listdir('/usr/lib/lxc/templates')

    if path:
        for line in path:
                templates.append(line.replace('lxc-', ''))

    return sorted(templates)


def check_version():
    '''
    returns latest LWP version (dict with current and latest)
    '''
    try:
        version = subprocess.check_output('git describe --tags', shell=True)
    except:
        version = open('version').read()[0:-1]
    return {'current': version}


def get_net_settings():
    '''
    returns a dict of all known settings for LXC networking
    '''
    filename = '/etc/default/lxc-net'
    if not file_exist(filename):
        filename = '/etc/default/lxc'
    if not file_exist(filename):
        return False
    if check_ubuntu() == "unknown":
        raise LxcConfigFileNotComplete('This is not a Ubuntu distro ! Check if all config params are set in /etc/default/lxc')
    config = ConfigParser.SafeConfigParser()
    cfg = {}
    config.readfp(FakeSection(open(filename)))
    cfg['use'] = config.get('DEFAULT', 'USE_LXC_BRIDGE').strip('"')
    cfg['bridge'] = config.get('DEFAULT', 'LXC_BRIDGE').strip('"')
    cfg['address'] = config.get('DEFAULT', 'LXC_ADDR').strip('"')
    cfg['netmask'] = config.get('DEFAULT', 'LXC_NETMASK').strip('"')
    cfg['network'] = config.get('DEFAULT', 'LXC_NETWORK').strip('"')
    cfg['range'] = config.get('DEFAULT', 'LXC_DHCP_RANGE').strip('"')
    cfg['max'] = config.get('DEFAULT', 'LXC_DHCP_MAX').strip('"')
    return cfg


def get_container_settings(name):
    '''
    returns a dict of all utils settings for a container
    '''
    filename = '/var/lib/lxc/%s/config' % name
    if not file_exist(filename):
        return False
    config = ConfigParser.SafeConfigParser()
    cfg = {}
    config.readfp(FakeSection(open(filename)))
    try:
        cfg['type'] = config.get('DEFAULT', cgroup['type'])
    except ConfigParser.NoOptionError:
        cfg['type'] = ''
    try:
        cfg['link'] = config.get('DEFAULT', cgroup['link'])
    except ConfigParser.NoOptionError:
        cfg['link'] = ''
    try:
        cfg['flags'] = config.get('DEFAULT', cgroup['flags'])
    except ConfigParser.NoOptionError:
        cfg['flags'] = ''
    try:
        cfg['hwaddr'] = config.get('DEFAULT', cgroup['hwaddr'])
    except ConfigParser.NoOptionError:
        cfg['hwaddr'] = ''
    try:
        cfg['rootfs'] = config.get('DEFAULT', cgroup['rootfs'])
    except ConfigParser.NoOptionError:
        cfg['rootfs'] = ''
    try:
        cfg['utsname'] = config.get('DEFAULT', cgroup['utsname'])
    except ConfigParser.NoOptionError:
        cfg['utsname'] = ''
    try:
        cfg['arch'] = config.get('DEFAULT', cgroup['arch'])
    except ConfigParser.NoOptionError:
        cfg['arch'] = ''
    try:
        cfg['ipv4'] = config.get('DEFAULT', cgroup['ipv4'])
    except ConfigParser.NoOptionError:
        cmd = ['lxc-ls --fancy --fancy-format name,ipv4|grep \'^%s \'|egrep -o \'[0-9]+\.[0-9]+\.[0-9]+\.[0-9]+\'' % name]
        try:
            cfg['ipv4'] = subprocess.check_output(cmd, shell=True)
            push_config_value(cgroup['ipv4'], cfg['ipv4'], name)
        except:
            cfg['ipv4'] = ''
    try:
        cfg['memlimit'] = re.sub(r'[a-zA-Z]', '', config.get('DEFAULT', cgroup['memlimit']))
    except ConfigParser.NoOptionError:
        cfg['memlimit'] = ''
    try:
        cfg['swlimit'] = re.sub(r'[a-zA-Z]', '', config.get('DEFAULT', cgroup['swlimit']))
    except ConfigParser.NoOptionError:
        cfg['swlimit'] = ''
    try:
        cfg['cpus'] = config.get('DEFAULT', cgroup['cpus'])
    except ConfigParser.NoOptionError:
        cfg['cpus'] = ''
    try:
        cfg['shares'] = config.get('DEFAULT', cgroup['shares'])
    except ConfigParser.NoOptionError:
        cfg['shares'] = ''
    try:
        cfg['auto'] = True if config.get('DEFAULT', cgroup['auto']) else False
    except ConfigParser.NoOptionError:
        cfg['auto'] = False

    return cfg


def push_net_value(key, value, filename='/etc/default/lxc'):
    '''
    replace a var in the lxc-net config file
    '''
    if filename:
        config = ConfigParser.RawConfigParser()
        config.readfp(FakeSection(open(filename)))
        if not value:
            config.remove_option('DEFAULT', key)
        else:
            config.set('DEFAULT', key, value)

        with open(filename, 'wb') as configfile:
            config.write(configfile)

        DelSection(filename=filename)

        load = open(filename, 'r')
        read = load.readlines()
        load.close()
        i = 0
        while i < len(read):
            if ' = ' in read[i]:
                split = read[i].split(' = ')
                split[1] = split[1].strip('\n')
                if '\"' in split[1]:
                    read[i] = '%s=%s\n' % (split[0].upper(), split[1])
                else:
                    read[i] = '%s=\"%s\"\n' % (split[0].upper(), split[1])
            i += 1
        load = open(filename, 'w')
        load.writelines(read)
        load.close()


def push_config_value(key, value, container=None):
    '''
    replace a var in a container config file
    '''

    def save_cgroup_devices(filename=None):
        '''
        returns multiple values (lxc.cgroup.devices.deny and lxc.cgroup.devices.allow) in a list.
        because ConfigParser cannot make this...
        '''
        if filename:
            values = []
            i = 0

            load = open(filename, 'r')
            read = load.readlines()
            load.close()

            while i < len(read):
                if not read[i].startswith('#') and re.match('lxc.cgroup.devices.deny|lxc.cgroup.devices.allow', read[i]):
                    values.append(read[i])
                i += 1
            return values

    if container:
        filename = '/var/lib/lxc/%s/config' % container
        save = save_cgroup_devices(filename=filename)

        config = ConfigParser.RawConfigParser()
        config.readfp(FakeSection(open(filename)))
        if not value:
            config.remove_option('DEFAULT', key)
        elif key == cgroup['memlimit'] or key == cgroup['swlimit'] and value is not False:
            config.set('DEFAULT', key, '%sM' % value)
        else:
            config.set('DEFAULT', key, value)

        # Bugfix (can't duplicate keys with config parser)
        if config.has_option('DEFAULT', cgroup['deny']) or config.has_option('DEFAULT', cgroup['allow']):
            config.remove_option('DEFAULT', cgroup['deny'])
            config.remove_option('DEFAULT', cgroup['allow'])

        with open(filename, 'wb') as configfile:
            config.write(configfile)

        DelSection(filename=filename)

        with open(filename, "a") as configfile:
            configfile.writelines(save)


def net_restart():
    '''
    restarts LXC networking
    '''
    cmd = ['/usr/sbin/service lxc-net restart']
    try:
        subprocess.check_call(cmd, shell=True)
        return 0
    except CalledProcessError:
        return 1
